components/webext-storage/src/sync/mod.rs

```diff
@@ -73,29 +94,81 @@ fn merge(mut other: JsonMap, mut ours: JsonMap, parent: Option<JsonMap>) -> Inco
                 "merge: key {} doesn't occur in parent - copying from incoming",
                 key
             );
+            changes.push(StorageValueChange {
+                key: key.clone(),
+                old_value: None,
+                new_value: Some(incoming_value.clone()),
+            });
             ours.insert(key, incoming_value);
         }
     } else {
         // No parent. Server wins. Overwrite every key in ours with
         // the corresponding value in other.
         log::trace!("merge: no parent - copying all keys from incoming");
         for (key, incoming_value) in other.into_iter() {
+            let old_value = ours.remove(&key);
+            let new_value = Some(incoming_value.clone());
+            if old_value != new_value {
+                changes.push(StorageValueChange {
+                    key: key.clone(),
+                    old_value,
+                    new_value,
+                });
+            }
             ours.insert(key, incoming_value);
         }
     }
 
     if ours == old_incoming {
-        IncomingAction::TakeRemote { data: old_incoming }
+        IncomingAction::TakeRemote {
+            data: old_incoming,
+            changes,
+        }
     } else {
-        IncomingAction::Merge { data: ours }
+        IncomingAction::Merge {
+            data: ours,
+            changes,
+        }
     }
 }
 
-fn remove_matching_keys(mut ours: JsonMap, blacklist: &JsonMap) -> JsonMap {
+fn remove_matching_keys(mut ours: JsonMap, blacklist: &JsonMap) -> (JsonMap, StorageChanges) {
+    let mut changes = StorageChanges::with_capacity(blacklist.len());
     for key in blacklist.keys() {
-        ours.remove(key);
+        if let Some(old_value) = ours.remove(key) {
+            changes.push(StorageValueChange {
+                key: key.clone(),
+                old_value: Some(old_value),
+                new_value: None,
+            });
+        }
     }
-    ours
+    (ours, changes)
+}
+
+// Fetches the applied changes we stashed in the storage_sync_applied table.
+// We don't convert to JSON because they need to be passed back to the browser
+// as strings anyway. However, we do return one string per extension rather than
+// a massive array.
```

Comment:
s/array/string/

Reply:
I went ahead and changed this to return a `Vec<SyncedExtensionChanges>` with separate fields for the `ext_id` and `changes`, like we talked about—because Desktop is going to need the `ext_id`, anyway, but we can pass changes through as they are.